// @rsocket/websocket-client/dist/WebsocketDuplexConnection.js
"use strict";
/*
 * Copyright 2021-2022 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebsocketDuplexConnection = void 0;
const core_1 = require("@rsocket/core");
class WebsocketDuplexConnection extends core_1.Deferred {
    constructor(websocket, deserializer, multiplexerDemultiplexerFactory) {
        super();
        this.websocket = websocket;
        this.deserializer = deserializer;
        this.handleClosed = (e) => {
            this.close(new Error(e.reason || "WebsocketDuplexConnection: Socket closed unexpectedly."));
        };
        this.handleError = (e) => {
            this.close(e.error);
        };
        this.handleMessage = (message) => {
            try {
                const buffer = Buffer.from(message.data);
                const frame = this.deserializer.deserializeFrame(buffer);
                this.multiplexerDemultiplexer.handle(frame);
            }
            catch (error) {
                this.close(error);
            }
        };
        websocket.addEventListener("close", this.handleClosed);
        websocket.addEventListener("error", this.handleError);
        websocket.addEventListener("message", this.handleMessage);
        this.multiplexerDemultiplexer = multiplexerDemultiplexerFactory(this);
    }
    get availability() {
        return this.done ? 0 : 1;
    }
    close(error) {
        if (this.done) {
            super.close(error);
            return;
        }
        this.websocket.removeEventListener("close", this.handleClosed);
        this.websocket.removeEventListener("error", this.handleError);
        this.websocket.removeEventListener("message", this.handleMessage);
        this.websocket.close();
        delete this.websocket;
        super.close(error);
    }
    send(frame) {
        if (this.done) {
            return;
        }
        const buffer = (0, core_1.serializeFrame)(frame);
        this.websocket.send(buffer);
    }
}
exports.WebsocketDuplexConnection = WebsocketDuplexConnection;
