// @rsocket/core/dist/Errors.js
"use strict";
/*
 * Copyright 2021-2022 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorCodes = exports.RSocketError = void 0;
class RSocketError extends Error {
    constructor(code, message) {
        super(message);
        this.code = code;
    }
}
exports.RSocketError = RSocketError;
var ErrorCodes;
(function (ErrorCodes) {
    ErrorCodes[ErrorCodes["RESERVED"] = 0] = "RESERVED";
    ErrorCodes[ErrorCodes["INVALID_SETUP"] = 1] = "INVALID_SETUP";
    ErrorCodes[ErrorCodes["UNSUPPORTED_SETUP"] = 2] = "UNSUPPORTED_SETUP";
    ErrorCodes[ErrorCodes["REJECTED_SETUP"] = 3] = "REJECTED_SETUP";
    ErrorCodes[ErrorCodes["REJECTED_RESUME"] = 4] = "REJECTED_RESUME";
    ErrorCodes[ErrorCodes["CONNECTION_CLOSE"] = 258] = "CONNECTION_CLOSE";
    ErrorCodes[ErrorCodes["CONNECTION_ERROR"] = 257] = "CONNECTION_ERROR";
    ErrorCodes[ErrorCodes["APPLICATION_ERROR"] = 513] = "APPLICATION_ERROR";
    ErrorCodes[ErrorCodes["REJECTED"] = 514] = "REJECTED";
    ErrorCodes[ErrorCodes["CANCELED"] = 515] = "CANCELED";
    ErrorCodes[ErrorCodes["INVALID"] = 516] = "INVALID";
    ErrorCodes[ErrorCodes["RESERVED_EXTENSION"] = 4294967295] = "RESERVED_EXTENSION";
})(ErrorCodes = exports.ErrorCodes || (exports.ErrorCodes = {}));
