// @rsocket/core/dist/Resume.js
"use strict";
/*
 * Copyright 2021-2022 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrameStore = void 0;
const _1 = require(".");
const Codecs_1 = require("./Codecs");
class FrameStore {
    constructor() {
        this.storedFrames = [];
        this._lastReceivedFramePosition = 0;
        this._firstAvailableFramePosition = 0;
        this._lastSentFramePosition = 0;
    }
    get lastReceivedFramePosition() {
        return this._lastReceivedFramePosition;
    }
    get firstAvailableFramePosition() {
        return this._firstAvailableFramePosition;
    }
    get lastSentFramePosition() {
        return this._lastSentFramePosition;
    }
    store(frame) {
        this._lastSentFramePosition += (0, Codecs_1.sizeOfFrame)(frame);
        this.storedFrames.push(frame);
    }
    record(frame) {
        this._lastReceivedFramePosition += (0, Codecs_1.sizeOfFrame)(frame);
    }
    dropTo(lastReceivedPosition) {
        let bytesToDrop = lastReceivedPosition - this._firstAvailableFramePosition;
        while (bytesToDrop > 0 && this.storedFrames.length > 0) {
            const storedFrame = this.storedFrames.shift();
            bytesToDrop -= (0, Codecs_1.sizeOfFrame)(storedFrame);
        }
        if (bytesToDrop !== 0) {
            throw new _1.RSocketError(_1.ErrorCodes.CONNECTION_ERROR, `State inconsistency. Expected bytes to drop ${lastReceivedPosition - this._firstAvailableFramePosition} but actual ${bytesToDrop}`);
        }
        this._firstAvailableFramePosition = lastReceivedPosition;
    }
    drain(consumer) {
        for (const frame of this.storedFrames) {
            consumer(frame);
        }
    }
}
exports.FrameStore = FrameStore;
