// @rsocket/core/dist/RSocketConnector.js
"use strict";
/*
 * Copyright 2021-2022 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RSocketConnector = void 0;
const ClientServerMultiplexerDemultiplexer_1 = require("./ClientServerMultiplexerDemultiplexer");
const Frames_1 = require("./Frames");
const RSocketSupport_1 = require("./RSocketSupport");
const Resume_1 = require("./Resume");
class RSocketConnector {
    constructor(config) {
        this.config = config;
    }
    connect() {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
        return __awaiter(this, void 0, void 0, function* () {
            const config = this.config;
            const setupFrame = {
                type: Frames_1.FrameTypes.SETUP,
                dataMimeType: (_b = (_a = config.setup) === null || _a === void 0 ? void 0 : _a.dataMimeType) !== null && _b !== void 0 ? _b : "application/octet-stream",
                metadataMimeType: (_d = (_c = config.setup) === null || _c === void 0 ? void 0 : _c.metadataMimeType) !== null && _d !== void 0 ? _d : "application/octet-stream",
                keepAlive: (_f = (_e = config.setup) === null || _e === void 0 ? void 0 : _e.keepAlive) !== null && _f !== void 0 ? _f : 60000,
                lifetime: (_h = (_g = config.setup) === null || _g === void 0 ? void 0 : _g.lifetime) !== null && _h !== void 0 ? _h : 300000,
                metadata: (_k = (_j = config.setup) === null || _j === void 0 ? void 0 : _j.payload) === null || _k === void 0 ? void 0 : _k.metadata,
                data: (_m = (_l = config.setup) === null || _l === void 0 ? void 0 : _l.payload) === null || _m === void 0 ? void 0 : _m.data,
                resumeToken: (_p = (_o = config.resume) === null || _o === void 0 ? void 0 : _o.tokenGenerator()) !== null && _p !== void 0 ? _p : null,
                streamId: 0,
                majorVersion: 1,
                minorVersion: 0,
                flags: (((_r = (_q = config.setup) === null || _q === void 0 ? void 0 : _q.payload) === null || _r === void 0 ? void 0 : _r.metadata) ? Frames_1.Flags.METADATA : Frames_1.Flags.NONE) |
                    (config.lease ? Frames_1.Flags.LEASE : Frames_1.Flags.NONE) |
                    (config.resume ? Frames_1.Flags.RESUME_ENABLE : Frames_1.Flags.NONE),
            };
            const connection = yield config.transport.connect((outbound) => {
                return config.resume
                    ? new ClientServerMultiplexerDemultiplexer_1.ResumableClientServerInputMultiplexerDemultiplexer(ClientServerMultiplexerDemultiplexer_1.StreamIdGenerator.create(-1), outbound, outbound, new Resume_1.FrameStore(), // TODO: add size control
                    setupFrame.resumeToken.toString(), (self, frameStore) => __awaiter(this, void 0, void 0, function* () {
                        const multiplexerDemultiplexerProvider = (outbound) => {
                            outbound.send({
                                type: Frames_1.FrameTypes.RESUME,
                                streamId: 0,
                                flags: Frames_1.Flags.NONE,
                                clientPosition: frameStore.firstAvailableFramePosition,
                                serverPosition: frameStore.lastReceivedFramePosition,
                                majorVersion: setupFrame.minorVersion,
                                minorVersion: setupFrame.majorVersion,
                                resumeToken: setupFrame.resumeToken,
                            });
                            return new ClientServerMultiplexerDemultiplexer_1.ResumeOkAwaitingResumableClientServerInputMultiplexerDemultiplexer(outbound, outbound, self);
                        };
                        let reconnectionAttempts = -1;
                        const reconnector = () => {
                            reconnectionAttempts++;
                            return config.resume
                                .reconnectFunction(reconnectionAttempts)
                                .then(() => config.transport
                                .connect(multiplexerDemultiplexerProvider)
                                .catch(reconnector));
                        };
                        yield reconnector();
                    }))
                    : new ClientServerMultiplexerDemultiplexer_1.ClientServerInputMultiplexerDemultiplexer(ClientServerMultiplexerDemultiplexer_1.StreamIdGenerator.create(-1), outbound, outbound);
            });
            const keepAliveSender = new RSocketSupport_1.KeepAliveSender(connection.multiplexerDemultiplexer.connectionOutbound, setupFrame.keepAlive);
            const keepAliveHandler = new RSocketSupport_1.KeepAliveHandler(connection, setupFrame.lifetime);
            const leaseHandler = config.lease
                ? new RSocketSupport_1.LeaseHandler((_s = config.lease.maxPendingRequests) !== null && _s !== void 0 ? _s : 256, connection.multiplexerDemultiplexer)
                : undefined;
            const responder = (_t = config.responder) !== null && _t !== void 0 ? _t : {};
            const connectionFrameHandler = new RSocketSupport_1.DefaultConnectionFrameHandler(connection, keepAliveHandler, keepAliveSender, leaseHandler, responder);
            const streamsHandler = new RSocketSupport_1.DefaultStreamRequestHandler(responder, 0);
            connection.onClose((e) => {
                keepAliveSender.close();
                keepAliveHandler.close();
                connectionFrameHandler.close(e);
            });
            connection.multiplexerDemultiplexer.connectionInbound(connectionFrameHandler);
            connection.multiplexerDemultiplexer.handleRequestStream(streamsHandler);
            connection.multiplexerDemultiplexer.connectionOutbound.send(setupFrame);
            keepAliveHandler.start();
            keepAliveSender.start();
            return new RSocketSupport_1.RSocketRequester(connection, (_v = (_u = config.fragmentation) === null || _u === void 0 ? void 0 : _u.maxOutboundFragmentSize) !== null && _v !== void 0 ? _v : 0, leaseHandler);
        });
    }
}
exports.RSocketConnector = RSocketConnector;
